File: src/KinectForEllai/KinectForEllai/WebSocketHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Web.Script.Serialization;
using WebSocketSharp;

namespace KinectForEllai
{
    class WebSocketHandler
    {
        private readonly WebSocket _webSocket;

        private readonly String _channel;

        private readonly JavaScriptSerializer _js = new JavaScriptSerializer();

        public WebSocketHandler(String host, String channel)
        {
            this._channel = channel;
            _webSocket = new WebSocket("ws://" + host);

            _webSocket.OnError += (sender, e) => {
               Console.Error.WriteLine("WebSocketHandler: " + e.Message);
            };

            _webSocket.OnClose += (sender, e) => {
               Console.Error.WriteLine("WebSocketHandler: " + e.Reason); 
            };

            Connect();
        }

        public String Send(object obj)
        {
            var message = new { channel = this._channel, payload = obj };
            String data = _js.Serialize(message);

            if (_webSocket.IsAlive)
            {
                _webSocket.Send(data);
            }
            else
            {
                Connect();
            }

            return data;
        }

        public void Disconnect()
        {
            _webSocket.Close();
        }

        private void Connect()
        {
            _webSocket.Connect();
        }

    }
}

File: src/OutputForEllai/js/websockethandler.js
/**
 * Websocket helper module
 */
var WebsocketHandler = function (host, channel) {
    'use strict';
    
    var socket = null;
    
    /**
     * Connect or reconnect to host
     */
    function connect() {
        socket = new WebSocket(host, "protocolOne");

        /**
         * Log connections
         */
        socket.onopen = function (event) {
            console.log("WebsocketHandler: Connected");
        };

        /**
         * Log errors
         */
        socket.onerror = function (event) {
            console.log("WebsocketHandler: ERROR ", event);
        }
        
        socket.onclose = function(event) {
            console.log("WebsocketHandler: Reconnecting");
            setTimeout(function(){
                connect();
            }, 500);
        }
        
        /**
         * Disect incoming messages into a channel and payload and
         * pass on to the delegate/callback function.
         */
        socket.onmessage = function (event) {
            var message = JSON.parse(event.data);

            if (ret.onMessage) {
                ret.onMessage(message.channel, message.payload);
            }
        }
    }
    
    /**
     * Disconnect a websocket connection
     */
    function disconnect() {
        socket.close();
    }

    /**
     * Send an object on the given channel.
     *
     * @param {object} payload  the object to send
     */
    function send(payload) {
        var message = {
            channel: channel,
            payload: payload
        };
        socket.send(JSON.stringify(message));
    }
    
    /**
     * Public API
     */
    var ret = {
        onMessage: null,
        send: send,
        disconnect: disconnect
    }
    
    connect();
    return ret;
};
